{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MetaNovo: probabilistic database filtration for accurate MS/MS peptide identification "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Matthys Potgieter, Andrew J.M. Nel, David Tabb, Jonathan Blackburn, Nicola Mulder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import shutil\n",
    "import scipy.stats\n",
    "import scikit_posthocs as ph\n",
    "import numpy as np\n",
    "import Bio\n",
    "import scipy.stats\n",
    "import scikit_posthocs as ph\n",
    "import numpy as np\n",
    "import matplotlib_venn\n",
    "from matplotlib_venn import venn3, venn2\n",
    "import matplotlib.pyplot as plt\n",
    "from pylab import *\n",
    "import matplotlib.pyplot as plt\n",
    "from pylab import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. MetaPro-IQ comparison"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.1 Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16  Raw file                                           Total\n",
       "    MS                                                198071\n",
       "    MS/MS                                             483091\n",
       "    MS/MS Submitted                                   523851\n",
       "    MS/MS Submitted (SIL)                             442331\n",
       "    MS/MS Submitted (ISO)                                  0\n",
       "    MS/MS Submitted (PEAK)                             81520\n",
       "    MS/MS Identified                                  140784\n",
       "    MS/MS Identified (SIL)                            136963\n",
       "    MS/MS Identified (ISO)                                 0\n",
       "    MS/MS Identified (PEAK)                             3821\n",
       "    MS/MS Identified [%]                               26.87\n",
       "    MS/MS Identified (SIL) [%]                         30.96\n",
       "    MS/MS Identified (PEAK) [%]                         4.69\n",
       "    Peptide Sequences Identified                       58473\n",
       "    Peaks                                        2.05843e+07\n",
       "    Isotope Patterns                             1.93684e+06\n",
       "    Isotope Patterns Sequenced                        394538\n",
       "    Isotope Patterns Sequenced (z>1)                  390720\n",
       "    Isotope Patterns Sequenced [%]                     20.37\n",
       "    Isotope Patterns Sequenced (z>1) [%]               23.56\n",
       "    Isotope Patterns Repeatedly Sequenced              36780\n",
       "    Isotope Patterns Repeatedly Sequenced [%]           9.32\n",
       "    Av. Absolute Mass Deviation [ppm]                0.62746\n",
       "    Mass Standard Deviation [ppm]                    0.89653\n",
       "    Av. Absolute Mass Deviation [mDa]                0.44035\n",
       "    Mass Standard Deviation [mDa]                    0.65842\n",
       "dtype: object"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('/home/thys/metanovo/metanovo_igc_txt/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16  Raw file                                           Total\n",
       "    MS                                                198071\n",
       "    MS/MS                                             483091\n",
       "    MS/MS Submitted                                   523851\n",
       "    MS/MS Submitted (SIL)                             442331\n",
       "    MS/MS Submitted (ISO)                                  0\n",
       "    MS/MS Submitted (PEAK)                             81520\n",
       "    MS/MS Identified                                  150135\n",
       "    MS/MS Identified (SIL)                            145957\n",
       "    MS/MS Identified (ISO)                                 0\n",
       "    MS/MS Identified (PEAK)                             4178\n",
       "    MS/MS Identified [%]                               28.66\n",
       "    MS/MS Identified (SIL) [%]                            33\n",
       "    MS/MS Identified (PEAK) [%]                         5.13\n",
       "    Peptide Sequences Identified                       62930\n",
       "    Peaks                                        2.05843e+07\n",
       "    Isotope Patterns                             1.93684e+06\n",
       "    Isotope Patterns Sequenced                        394538\n",
       "    Isotope Patterns Sequenced (z>1)                  390720\n",
       "    Isotope Patterns Sequenced [%]                     20.37\n",
       "    Isotope Patterns Sequenced (z>1) [%]               23.56\n",
       "    Isotope Patterns Repeatedly Sequenced              36780\n",
       "    Isotope Patterns Repeatedly Sequenced [%]           9.32\n",
       "    Av. Absolute Mass Deviation [ppm]                0.62399\n",
       "    Mass Standard Deviation [ppm]                    0.89449\n",
       "    Av. Absolute Mass Deviation [mDa]                0.43352\n",
       "    Mass Standard Deviation [mDa]                    0.65007\n",
       "dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('/home/thys/metanovo/metagenome_metanovo_txt/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16  Raw file                                           Total\n",
       "    MS                                                198071\n",
       "    MS/MS                                             483091\n",
       "    MS/MS Submitted                                   523851\n",
       "    MS/MS Submitted (SIL)                             442331\n",
       "    MS/MS Submitted (ISO)                                  0\n",
       "    MS/MS Submitted (PEAK)                             81520\n",
       "    MS/MS Identified                                  190106\n",
       "    MS/MS Identified (SIL)                            185213\n",
       "    MS/MS Identified (ISO)                                 0\n",
       "    MS/MS Identified (PEAK)                             4893\n",
       "    MS/MS Identified [%]                               36.29\n",
       "    MS/MS Identified (SIL) [%]                         41.87\n",
       "    MS/MS Identified (PEAK) [%]                            6\n",
       "    Peptide Sequences Identified                       69499\n",
       "    Peaks                                        2.05843e+07\n",
       "    Isotope Patterns                             1.93684e+06\n",
       "    Isotope Patterns Sequenced                        394538\n",
       "    Isotope Patterns Sequenced (z>1)                  390720\n",
       "    Isotope Patterns Sequenced [%]                     20.37\n",
       "    Isotope Patterns Sequenced (z>1) [%]               23.56\n",
       "    Isotope Patterns Repeatedly Sequenced              36780\n",
       "    Isotope Patterns Repeatedly Sequenced [%]           9.32\n",
       "    Av. Absolute Mass Deviation [ppm]                0.61329\n",
       "    Mass Standard Deviation [ppm]                    0.88381\n",
       "    Av. Absolute Mass Deviation [mDa]                0.43232\n",
       "    Mass Standard Deviation [mDa]                    0.65211\n",
       "dtype: object"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('/home/thys/metanovo/txt_metanovo/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32  Raw file                                           Total\n",
       "    MS                                                186376\n",
       "    MS/MS                                             646329\n",
       "    MS3                                                    0\n",
       "    MS/MS Submitted                                   753677\n",
       "    MS/MS Submitted (SIL)                             538979\n",
       "    MS/MS Submitted (ISO)                                  0\n",
       "    MS/MS Submitted (PEAK)                            214698\n",
       "    MS/MS Identified                                  283256\n",
       "    MS/MS Identified (SIL)                            273112\n",
       "    MS/MS Identified (ISO)                                 0\n",
       "    MS/MS Identified (PEAK)                            10144\n",
       "    MS/MS Identified [%]                               37.58\n",
       "    MS/MS Identified (SIL) [%]                         50.67\n",
       "    MS/MS Identified (PEAK) [%]                         4.72\n",
       "    Peptide Sequences Identified                       25250\n",
       "    Peaks                                        2.26897e+07\n",
       "    Isotope Patterns                             1.86192e+06\n",
       "    Isotope Patterns Sequenced                        346849\n",
       "    Isotope Patterns Sequenced (z>1)                  335525\n",
       "    Isotope Patterns Sequenced [%]                     18.63\n",
       "    Isotope Patterns Sequenced (z>1) [%]               23.08\n",
       "    Isotope Patterns Repeatedly Sequenced             125518\n",
       "    Isotope Patterns Repeatedly Sequenced [%]          36.19\n",
       "    Av. Absolute Mass Deviation [ppm]                0.79609\n",
       "    Mass Standard Deviation [ppm]                     1.1122\n",
       "    Av. Absolute Mass Deviation [mDa]                0.53742\n",
       "    Mass Standard Deviation [mDa]                    0.78763\n",
       "dtype: object"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('/home/thys/metanovo/cns_reference_txt/summary.txt',sep='\\t')\n",
    "summary[32:].stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('cns_metanovo_txt/summary.txt',sep='\\t')\n",
    "summary[32:].stack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis: metanovo - UniProt\n",
      "metanovo - UniProt: Total peptides:  69499\n",
      "metanovo - UniProt: Total target peptides:  69314\n",
      "metanovo - UniProt: proteins:  15374\n",
      "metanovo - UniProt: Total target proteins:  15204\n",
      "\n",
      "Analysis: metanovo - igc\n",
      "metanovo - igc: Total peptides:  58473\n",
      "metanovo - igc: Total target peptides:  58307\n",
      "metanovo - igc: proteins:  14004\n",
      "metanovo - igc: Total target proteins:  13852\n",
      "\n",
      "Analysis: metanovo - metagenome\n",
      "metanovo - metagenome: Total peptides:  62930\n",
      "metanovo - metagenome: Total target peptides:  62774\n",
      "metanovo - metagenome: proteins:  13076\n",
      "metanovo - metagenome: Total target proteins:  12933\n",
      "\n",
      "Analysis: metaproiq - igc\n",
      "metaproiq - igc: Total peptides:  69051\n",
      "metaproiq - igc: Total target peptides:  68828\n",
      "metaproiq - igc: proteins:  19742\n",
      "metaproiq - igc: Total target proteins:  19533\n",
      "\n",
      "Analysis: metaproiq - metagenome\n",
      "metaproiq - metagenome: Total peptides:  71596\n",
      "metaproiq - metagenome: Total target peptides:  71390\n",
      "metaproiq - metagenome: proteins:  16956\n",
      "metaproiq - metagenome: Total target proteins:  16777\n",
      "\n"
     ]
    }
   ],
   "source": [
    "folders = { 'metanovo - UniProt':'txt_metanovo',\n",
    "            'metanovo - igc':'metanovo_igc_txt',\n",
    "            'metanovo - metagenome':'metagenome_metanovo_txt',\n",
    "            'metaproiq - igc' :'txt_workflow',\n",
    "            'metaproiq - metagenome':'txt_matched_metagenome'}\n",
    "\n",
    "order = ['metanovo - UniProt', 'metanovo - igc', 'metanovo - metagenome' , 'metaproiq - igc' , 'metaproiq - metagenome']\n",
    "def process_txt(path, name):\n",
    "    print('Analysis: {}'.format(name))\n",
    "    result = {}\n",
    "    peptides = pd.read_csv('{}/peptides.txt'.format(path),sep=None, engine='python')\n",
    "    proteins = pd.read_csv('{}/proteinGroups.txt'.format(path),sep=None, engine='python')\n",
    "    print('{}: Total peptides: '.format(name),len(peptides))\n",
    "    target_peptides = peptides[peptides['Reverse']!= '+']\n",
    "    reverse_peptides = peptides[peptides['Reverse'] == '+']\n",
    "    print('{}: Total target peptides: '.format(name),len(target_peptides))\n",
    "    print('{}: proteins: '.format(name),len(proteins))\n",
    "    target_proteins = proteins[proteins['Reverse']!= '+']\n",
    "    print('{}: Total target proteins: '.format(name),len(target_proteins))\n",
    "    peptide_sequences = set(target_peptides['Sequence'].tolist())\n",
    "    result['TargetPeptides'] = target_peptides\n",
    "    result['ReversePeptides'] = reverse_peptides\n",
    "\n",
    "    result['TargetProteins'] = target_proteins\n",
    "    reverse_peptide_sequences = reverse_peptides['Sequence'].tolist()\n",
    "    w = open('analysis/{}_peptide_list.txt'.format(name),'w')\n",
    "    w.write('\\n'.join(peptide_sequences))\n",
    "    w.close()\n",
    "    print()\n",
    "    return result\n",
    "\n",
    "results = {}\n",
    "for name in order:\n",
    "    path = folders[name]\n",
    "    res = process_txt(path, name)\n",
    "    results[name] = res\n",
    "    #print(res.keys())\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.2 Peptide and Protein Identification Bar Charts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "keys = list(results.keys())\n",
    "plt.clf()\n",
    "# Peptide counts per sample\n",
    "count_df = pd.DataFrame()\n",
    "for name in order:\n",
    "    target_peptides = results[name]['TargetPeptides']\n",
    "    for col in target_peptides.columns:\n",
    "        if col.startswith('Experiment'):\n",
    "            seqs = set(target_peptides[(target_peptides[col] > 0)]['Sequence'].tolist())\n",
    "            count_df.loc[col.split()[1], name] = int(len(seqs))\n",
    "ax1 = count_df.plot(kind='bar', rot=1)\n",
    "ax1.set_title(\"Number of peptides\")\n",
    "ax1.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "fig = ax1.get_figure()\n",
    "fig.savefig('analysis/peptide_id_bar.png', bbox_inches='tight')\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# Protein counts per sample\n",
    "prot_count = pd.DataFrame()\n",
    "for name in order:\n",
    "    target_proteins = results[name]['TargetProteins']\n",
    "    for col in target_proteins.columns:\n",
    "        if col.startswith('MS/MS Count '):\n",
    "            ids = set(target_proteins[(target_proteins[col] > 0)]['id'].tolist())\n",
    "            prot_count.loc[col.split()[-1], name] = int(len(ids))\n",
    "ax2 = prot_count.plot(kind='bar', rot=1)\n",
    "ax2.set_title(\"Number of protein groups\")\n",
    "ax2.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "fig = ax2.get_figure()\n",
    "fig.savefig('analysis/protein_id_bar.png', bbox_inches='tight')\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.3 PEP Score boxplots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "## Peptide Sets ##\n",
    "##################\n",
    "all_peps = set()\n",
    "keys = list(results.keys())\n",
    "names = []\n",
    "pep_scores = []\n",
    "for key in order:\n",
    "    peps = set(results[key]['TargetPeptides']['Sequence'])\n",
    "    all_peps.update(peps)\n",
    "# Get common peptides to all runs\n",
    "intersect = all_peps.copy()\n",
    "for key in order:\n",
    "    peps = set(results[key]['TargetPeptides']['Sequence'])\n",
    "    intersect = intersect & peps\n",
    "# Get esclusive peptides for all runs\n",
    "for key in order:\n",
    "    peps = results[key]['TargetPeptides']\n",
    "    rpeps = results[key]['ReversePeptides']\n",
    "    exclusive = set(peps['Sequence'].tolist()).copy()\n",
    "    for qkey in keys:\n",
    "        if not qkey == key:\n",
    "            qpeps = results[qkey]['TargetPeptides']\n",
    "            qpeps = set(qpeps['Sequence'].tolist())\n",
    "            exclusive -= qpeps\n",
    "    print('{} exclusive: '.format(key), len(exclusive))\n",
    "    common_pep = peps[peps['Sequence'].isin(intersect)]['PEP'].tolist()\n",
    "    pep_scores.append(common_pep)\n",
    "    names.append(key + ': Shared')\n",
    "    exclusive_pep = peps[peps['Sequence'].isin(exclusive)]['PEP'].tolist()\n",
    "    pep_scores.append(exclusive_pep)\n",
    "    names.append(key + ': Only')\n",
    "    reverse_pep = rpeps['PEP'].tolist()\n",
    "    pep_scores.append(reverse_pep)\n",
    "    names.append(key + ': Reverse')\n",
    "\n",
    "#############\n",
    "## BOXPLOT ##\n",
    "#############\n",
    "\n",
    "colours = ['b','g','r','c','m','y','k']\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "# Create the boxplot\n",
    "bp = ax.boxplot(pep_scores, patch_artist=True, showfliers=False)\n",
    "## change outline color, fill color and linewidth of the boxes\n",
    "count = 0\n",
    "col_ind=0\n",
    "for box in bp['boxes']:\n",
    "    count += 1\n",
    "    # change outline color\n",
    "    box.set( color='#7570b3', linewidth=1)\n",
    "    # change fill color\n",
    "    box.set( facecolor = colours[col_ind] )\n",
    "    ## change color and linewidth of the whiskers\n",
    "    if count % 3 == 0:\n",
    "        col_ind +=1\n",
    "for whisker in bp['whiskers']:\n",
    "    whisker.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the caps\n",
    "for cap in bp['caps']:\n",
    "    cap.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the medians\n",
    "for median in bp['medians']:\n",
    "    median.set(color='#b2df8a', linewidth=2)\n",
    "    #median.set(linewidth=2)\n",
    "## change the style of fliers and their fill\n",
    "for flier in bp['fliers']:\n",
    "    flier.set(marker='.', color='#e7298a', alpha=0.5)\n",
    "## Custom x-axis labels\n",
    "ax.set_xticklabels(names, rotation=90)\n",
    "#ax.set_yticklabels('Posterior Error Probability (PEP)') \n",
    "ax.set_title('Peptide PEP Score distributions')\n",
    "## Remove top axes and right axes ticks\n",
    "ax.get_xaxis().tick_bottom()\n",
    "ax.get_yaxis().tick_left()\n",
    "fig.savefig('analysis/psm_pep_scores.png', bbox_inches='tight')\n",
    "plt.show()\n",
    "fig.clf()\n",
    "\n",
    "pep_df = pd.DataFrame()\n",
    "for val in range(len(names)):\n",
    "    pep_df.loc[names[val], 'Count'] = len(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - median'] = np.median(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - std. dev.'] = np.std(pep_scores[val])\n",
    "\n",
    "pep_df.to_csv('analysis/pep_scores.csv')\n",
    "pep_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.4 PEP Score Kruskal-Wallis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KruskalResult(statistic=261.7432734591897, pvalue=1.8866788880406473e-56)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metanovo - UniProt: Human</th>\n",
       "      <th>metanovo - UniProt: Other</th>\n",
       "      <th>metanovo - UniProt: Contaminant</th>\n",
       "      <th>metanovo - UniProt: Reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Human</th>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>7.074290e-51</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>3.375956e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Other</th>\n",
       "      <td>7.074290e-51</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.994544</td>\n",
       "      <td>6.978041e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Contaminant</th>\n",
       "      <td>2.475835e-06</td>\n",
       "      <td>9.945444e-01</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>1.437990e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Reverse</th>\n",
       "      <td>3.375956e-15</td>\n",
       "      <td>6.978041e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 metanovo - UniProt: Human  \\\n",
       "metanovo - UniProt: Human                    -1.000000e+00   \n",
       "metanovo - UniProt: Other                     7.074290e-51   \n",
       "metanovo - UniProt: Contaminant               2.475835e-06   \n",
       "metanovo - UniProt: Reverse                   3.375956e-15   \n",
       "\n",
       "                                 metanovo - UniProt: Other  \\\n",
       "metanovo - UniProt: Human                     7.074290e-51   \n",
       "metanovo - UniProt: Other                    -1.000000e+00   \n",
       "metanovo - UniProt: Contaminant               9.945444e-01   \n",
       "metanovo - UniProt: Reverse                   6.978041e-08   \n",
       "\n",
       "                                 metanovo - UniProt: Contaminant  \\\n",
       "metanovo - UniProt: Human                               0.000002   \n",
       "metanovo - UniProt: Other                               0.994544   \n",
       "metanovo - UniProt: Contaminant                        -1.000000   \n",
       "metanovo - UniProt: Reverse                             0.000001   \n",
       "\n",
       "                                 metanovo - UniProt: Reverse  \n",
       "metanovo - UniProt: Human                       3.375956e-15  \n",
       "metanovo - UniProt: Other                       6.978041e-08  \n",
       "metanovo - UniProt: Contaminant                 1.437990e-06  \n",
       "metanovo - UniProt: Reverse                    -1.000000e+00  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def list_kw_dunn(names, data, value, group, path):\n",
    "    colnames=names\n",
    "    kw = scipy.stats.kruskal(*data)\n",
    "    w = open(path + '/kw.txt', 'w')\n",
    "    w.write(str(kw))\n",
    "    print(kw)\n",
    "    w.close()\n",
    "    post_hoc = pd.DataFrame(ph.posthoc_dunn(data, p_adjust = 'fdr_bh'))\n",
    "    post_hoc.index = names\n",
    "    post_hoc.columns = names\n",
    "    post_hoc.to_csv(path + '/dunn_bh.csv')\n",
    "    return post_hoc\n",
    "post_hoc = list_kw_dunn(names, pep_scores, \"PEP\", \"Workflow\", 'analysis')\n",
    "post_hoc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.5 Venn Diagrams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "set1 = set(results['metanovo - UniProt']['TargetPeptides']['Sequence'])\n",
    "set2 = set(results['metaproiq - igc']['TargetPeptides']['Sequence'])\n",
    "set3 = set(results['metaproiq - metagenome']['TargetPeptides']['Sequence'])\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "fig = plt.figure()\n",
    "venn3([set1, set2, set3], ('MetaNovo:\\nUniProt', 'MetaPro-IQ:\\nintegrated gene catalog', 'MetaProIQ:\\nmatched metagenome'))\n",
    "fig.savefig('analysis/mli_venn1.png', bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "fig = plt.figure()\n",
    "venn2([set1, set2], ('MetaNovo:\\nUniProt', 'MetaPro-IQ:\\nintegrated gene catalog'))\n",
    "fig.savefig('analysis/mli_venn2.png', bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "set1 = set(results['metanovo - UniProt']['TargetPeptides']['Sequence'])\n",
    "set2 = set(results['metanovo - igc']['TargetPeptides']['Sequence'])\n",
    "set3 = set(results['metanovo - metagenome']['TargetPeptides']['Sequence'])\n",
    "\n",
    "fig = plt.figure()\n",
    "venn3([set1, set2, set3], ('MetaNovo:\\nUniProt', 'MetaNovo:\\nintegrated gene catalog', 'MetaNovo:\\nmatched metagenome'))\n",
    "fig.savefig('analysis/mli_venn3.png', bbox_inches='tight')\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "\n",
    "\n",
    "plt.show()\n",
    "\n",
    "set1 = set(results['metanovo - igc']['TargetPeptides']['Sequence'])\n",
    "set2 = set(results['metaproiq - igc']['TargetPeptides']['Sequence'])\n",
    "\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "fig = plt.figure()\n",
    "venn2([set1, set2], ('MetaNovo:\\nintegrated gene catalog', 'MetaPro-IQ:\\nintegrated gene catalog'))\n",
    "fig.savefig('analysis/mli_venn4.png', bbox_inches='tight')\n",
    "\n",
    "plt.show()\n",
    "\n",
    "set1 = set(results['metanovo - metagenome']['TargetPeptides']['Sequence'])\n",
    "set2 = set(results['metaproiq - metagenome']['TargetPeptides']['Sequence'])\n",
    "\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "fig = plt.figure()\n",
    "venn2([set1, set2], ('MetaNovo:\\nmatched metagenome', 'MetaPro-IQ:\\nmatched metagenome'))\n",
    "fig.savefig('analysis/mli_venn5.png', bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1.6 UniPept "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('All peptides found: ', len(all_peps))\n",
    "w = open('analysis/combined_peptide_set.txt','w')\n",
    "w.write('\\n'.join(all_peps))\n",
    "w.close()\n",
    "cmd = 'cat analysis/combined_peptide_set.txt | prot2pept | peptfilter | sort -u | unipept pept2lca -e -a > analysis/combined_pept2lca.csv'\n",
    "!{cmd}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "taxa = pd.read_csv('analysis/combined_pept2lca.csv')\n",
    "template = pd.DataFrame(pd.Series(list(all_peps)))\n",
    "template.rename(columns={0:'peptide'}, inplace=True)\n",
    "smapping = pd.merge(template, taxa, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "order = ['metanovo - UniProt', 'metanovo - igc', 'metanovo - metagenome' , 'metaproiq - igc' , 'metaproiq - metagenome']\n",
    "def plot_taxa( results, col , mapping):\n",
    "    keys = list(results.keys())\n",
    "    plt.clf()\n",
    "    # Peptide counts per sample\n",
    "    count_df = pd.DataFrame()\n",
    "    for name in order:\n",
    "        target_peptides = results[name]['TargetPeptides']\n",
    "        mapped = pd.merge(target_peptides, mapping, how='left', left_on='Sequence', right_on='peptide')\n",
    "        #mapped = mapped.replace(np.nan, 'Uncharacterized')\n",
    "        agg_cols = {'MS/MS Count': sum }\n",
    "        agg = mapped.groupby(mapped[col]).agg(agg_cols)\n",
    "        count_df[name] = agg['MS/MS Count']\n",
    "    count_df = count_df.replace(np.nan, 0)\n",
    "    count_df['Total'] = count_df.sum(axis=1)\n",
    "    for column in count_df.columns:\n",
    "        vals = count_df[count_df[column] > 0 ]\n",
    "        print(column, len(vals))\n",
    "        \n",
    "    count_df = count_df.sort_values('Total',ascending=False)[:15]\n",
    "    #print(count_df.head())\n",
    "    del count_df['Total']\n",
    "    t_df = count_df.transpose()    \n",
    "    ax1 = t_df.plot(kind='bar', rot=1, stacked=True)\n",
    "    ax1.set_title(\"MS/MS Count by UniPept pept2lca {}\".format(col))\n",
    "    ax1.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "    plt.xticks(rotation=90)\n",
    "    fig = ax1.get_figure()\n",
    "    fig.savefig('analysis/{}_bar.png'.format(col), bbox_inches='tight')\n",
    "    \n",
    "    plt.show()\n",
    "    #print(count_df)\n",
    "    return count_df\n",
    "\n",
    "count_df = plot_taxa(results, 'phylum_name', smapping)\n",
    "#plot_taxa('family_name')\n",
    "#count_df = plot_taxa('genus_name')\n",
    "count_df = plot_taxa(results, 'species_name', smapping)\n",
    "\n",
    "#count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_taxa_runs( data_results, col , run, mapping):\n",
    "    keys = list(data_results.keys())\n",
    "    plt.clf()\n",
    "    # Peptide counts per sample\n",
    "    count_df = pd.DataFrame()\n",
    "    target_peptides = data_results[run]['TargetPeptides']\n",
    "    for c in target_peptides.columns:\n",
    "            if c.startswith('Experiment '):\n",
    "                seqs = target_peptides[[ 'Sequence' , c ]]\n",
    "                #count_df.loc[col.split()[1], name] = int(len(seqs))\n",
    "                mapped = pd.merge(seqs, mapping, how='left', left_on='Sequence', right_on='peptide')\n",
    "                #mapped[col] = mapped[col].replace(np.nan, 'Uncharacterized')\n",
    "                agg_cols = {c: sum }\n",
    "                agg = mapped.groupby(mapped[col]).agg(agg_cols)\n",
    "                agg[c] = agg[c]#/agg[c].sum() * 100\n",
    "                count_df[c.split()[-1]] = agg[c]\n",
    "    count_df = count_df.replace(np.nan, 0)\n",
    "    count_df['Total'] = count_df.sum(axis=1)\n",
    "    count_df = count_df.sort_values('Total',ascending=False)[:15]\n",
    "    del count_df['Total']\n",
    "    t_df = count_df.transpose()    \n",
    "    ax1 = t_df.plot(kind='bar', rot=1, stacked=True)\n",
    "    ax1.set_title(\"MS/MS Count by UniPept pept2lca {}\".format(col))\n",
    "    ax1.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "    fig = ax1.get_figure()\n",
    "    fig.savefig('analysis/{}_bar.png'.format(col), bbox_inches='tight')\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.show()\n",
    "    return count_df\n",
    "\n",
    "print('MetaGenome')\n",
    "plot_taxa_runs(results, 'phylum_name', 'metaproiq - metagenome', smapping )\n",
    "print('IGC')\n",
    "plot_taxa_runs(results, 'phylum_name', 'metaproiq - igc', smapping )\n",
    "print('UniProt')\n",
    "plot_taxa_runs(results, 'phylum_name', 'metanovo - UniProt' , smapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('MetaGenome')\n",
    "plot_taxa_runs(results, 'species_name', 'metaproiq - metagenome', smapping )\n",
    "print('IGC')\n",
    "plot_taxa_runs(results, 'species_name', 'metaproiq - igc' ,smapping)\n",
    "print('UniProt')\n",
    "plot_taxa_runs(results, 'species_name', 'metanovo - UniProt' , smapping)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Human CNS Validation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.1 Preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('cns_reference_txt/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('cns_metanovo_txt/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "folders = {'metanovo - UniProt':'cns_metanovo_txt',\n",
    "           'Human Reference proteome':'cns_reference_txt'}\n",
    "\n",
    "human_results = {}\n",
    "order = list(folders.keys())\n",
    "for name in folders:\n",
    "    path = folders[name]\n",
    "    res = process_txt(path, name)\n",
    "    human_results[name] = res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.2 Peptide and protein identification bar charts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "keys = list(human_results.keys())\n",
    "plt.clf()\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "# Peptide counts per sample\n",
    "count_df = pd.DataFrame()\n",
    "for name in keys:\n",
    "    target_peptides = human_results[name]['TargetPeptides']\n",
    "    for col in target_peptides.columns:\n",
    "        if col.startswith('Experiment'):\n",
    "            seqs = set(target_peptides[(target_peptides[col] > 0)]['Sequence'].tolist())\n",
    "            count_df.loc[col.split()[1], name] = int(len(seqs))\n",
    "ax1 = count_df.plot(kind='bar', rot=1)\n",
    "ax1.set_title(\"Number of peptides\")\n",
    "ax1.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "fig = ax1.get_figure()\n",
    "\n",
    "fig.savefig('analysis/peptide_id_bar_cns.png', bbox_inches='tight')\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "#figure(num=None, figsize=(8, 6), dpi=120, facecolor='w', edgecolor='k')\n",
    "# Protein counts per sample\n",
    "prot_count = pd.DataFrame()\n",
    "for name in keys:\n",
    "    target_proteins = human_results[name]['TargetProteins']\n",
    "    for col in target_proteins.columns:\n",
    "        if col.startswith('MS/MS count '):\n",
    "            ids = set(target_proteins[(target_proteins[col] > 0)]['id'].tolist())\n",
    "            prot_count.loc[col.split()[-1], name] = int(len(ids))\n",
    "##prot_count['index'] = prot_count.index\n",
    "#prot_count['index'] = prot_count['index'].apply(lambda x : int(x.split(\"E\")[1]))\n",
    "#prot_count = prot_count.sort_values('index')\n",
    "#del prot_count['index']\n",
    "ax2 = prot_count.plot(kind='bar', rot=1)\n",
    "ax2.set_title(\"Number of protein groups\")\n",
    "ax2.legend(bbox_to_anchor=(1.05, 0), loc='lower left', borderaxespad=0.)\n",
    "fig = ax2.get_figure()\n",
    "fig.savefig('analysis/protein_id_bar_cns.png', bbox_inches='tight')\n",
    "plt.xticks(rotation=90)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.3 Peptide identification VENN diagrams"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hset1 = set(human_results['metanovo - UniProt']['TargetPeptides']['Sequence'])\n",
    "hset2 = set(human_results['Human Reference proteome']['TargetPeptides']['Sequence'])\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "\n",
    "\n",
    "venn2([hset1, hset2], ('MetaNovo:\\nUniProt', 'Rerence proteome'))\n",
    "fig.savefig('analysis/mli_venn_cns.png', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_cns_peps = hset1.union(hset2)\n",
    "print('All peptides found: ', len(all_cns_peps))\n",
    "w = open('analysis/combined_cns_peptide_set.txt','w')\n",
    "w.write('\\n'.join(all_cns_peps))\n",
    "w.close()\n",
    "cmd = 'cat analysis/combined_cns_peptide_set.txt | prot2pept | peptfilter | sort -u | unipept pept2lca -e -a > analysis/combined_cns_pept2lca.csv'\n",
    "!{cmd}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(hset1 & hset2) / len(hset1 | hset2) * 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_results.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "htaxa = pd.read_csv('analysis/combined_cns_pept2lca.csv')\n",
    "htemplate = pd.DataFrame(pd.Series(list(all_cns_peps)))\n",
    "htemplate.rename(columns={0:'peptide'}, inplace=True)\n",
    "hmapping = pd.merge(htemplate, htaxa, how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = plot_taxa(human_results, 'phylum_name',  hmapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = plot_taxa(human_results, 'species_name', hmapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('UniProt')\n",
    "res = plot_taxa_runs(human_results, 'species_name', 'metanovo - UniProt', hmapping )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print('Reference proteome')\n",
    "res = plot_taxa_runs(human_results, 'species_name', 'Human Reference proteome', hmapping )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.4 PEP Score analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "human_reference = list(SeqIO.parse(\"cns_reference_txt/uniprot-proteome%3AUP000005640.fasta\",\"fasta\"))\n",
    "seqs = '*'.join([str(i.seq) for i in human_reference])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scipy.stats\n",
    "import scikit_posthocs as ph\n",
    "import numpy as np\n",
    "import Bio\n",
    "\n",
    "##################\n",
    "## Peptide Sets ##\n",
    "##################\n",
    "\n",
    "all_human_peps = set()\n",
    "all_non_human_peps = set()\n",
    "\n",
    "keys = list(human_results.keys())\n",
    "names = []\n",
    "pep_scores = []\n",
    "for key in keys:\n",
    "    print(key)\n",
    "    peps = set(human_results[key]['TargetPeptides']['Sequence'])\n",
    "    for pep in peps:\n",
    "        if pep in seqs:\n",
    "            all_human_peps.add(pep)\n",
    "        else:\n",
    "            all_non_human_peps.add(pep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "contaminants = human_results['metanovo - UniProt']['TargetPeptides']\n",
    "contaminants = set(contaminants[contaminants['Potential contaminant'] == '+']['Sequence'].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_non_human_peps\n",
    "\n",
    "temp = human_results['metanovo - UniProt']['TargetPeptides']\n",
    "variants = temp[temp['Sequence'].isin(all_non_human_peps)]\n",
    "print(len(variants))\n",
    "variants = variants[variants['Potential contaminant'] != '+']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pep_scores = []\n",
    "names = []\n",
    "for key in keys:\n",
    "    peps = human_results[key]['TargetPeptides']\n",
    "    cpeps = peps[peps['Potential contaminant']=='+']\n",
    "    peps = peps[peps['Potential contaminant'] != '+']\n",
    "    #peps = peps[peps['PEP'] < 1]\n",
    "    rpeps = human_results[key]['ReversePeptides']\n",
    "    key_peps = set(peps['Sequence'].tolist()).copy()\n",
    "    non_human= set(key_peps)\n",
    "    human = set(key_peps)\n",
    "    for qkey in keys:\n",
    "        if not qkey == key:\n",
    "            non_human -= all_human_peps\n",
    "            human -= all_non_human_peps\n",
    "            \n",
    "    print('{} Human: '.format(key), len(human))\n",
    "    print('{} Non-human: '.format(key), len(non_human))\n",
    "        \n",
    "    human_pep = peps[peps['Sequence'].isin(human)]['PEP'].tolist()\n",
    "    pep_scores.append(human_pep)\n",
    "    names.append(key + ': Human')\n",
    "    \n",
    "    non_human_pep = peps[peps['Sequence'].isin(non_human)]['PEP'].tolist()\n",
    "    pep_scores.append(non_human_pep)\n",
    "    names.append(key + ': Other')\n",
    "    \n",
    "    contaminant_pep = cpeps['PEP'].tolist()\n",
    "    pep_scores.append(contaminant_pep)\n",
    "    names.append(key + ': Contaminant' )\n",
    "    \n",
    "    reverse_pep = rpeps['PEP'].tolist()\n",
    "    pep_scores.append(reverse_pep)\n",
    "    names.append(key + ': Reverse')\n",
    "\n",
    "#############\n",
    "## BOXPLOT ##\n",
    "#############\n",
    "\n",
    "colours = ['b','g','r','c','m','y','k']\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "# Create the boxplot\n",
    "bp = ax.boxplot(pep_scores, patch_artist=True, showfliers=False)\n",
    "## change outline color, fill color and linewidth of the boxes\n",
    "count = 0\n",
    "col_ind=0\n",
    "for box in bp['boxes']:\n",
    "    count += 1\n",
    "    # change outline color\n",
    "    box.set( color='#7570b3', linewidth=1)\n",
    "    # change fill color\n",
    "    box.set( facecolor = colours[col_ind] )\n",
    "    ## change color and linewidth of the whiskers\n",
    "    if count % 4 == 0:\n",
    "        col_ind +=1\n",
    "for whisker in bp['whiskers']:\n",
    "    whisker.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the caps\n",
    "for cap in bp['caps']:\n",
    "    cap.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the medians\n",
    "for median in bp['medians']:\n",
    "    median.set(color='#b2df8a', linewidth=2)\n",
    "    #median.set(linewidth=2)\n",
    "## change the style of fliers and their fill\n",
    "for flier in bp['fliers']:\n",
    "    flier.set(marker='.', color='#e7298a', alpha=0.5)\n",
    "## Custom x-axis labels\n",
    "ax.set_xticklabels(names, rotation=90)\n",
    "#ax.set_yticklabels('Posterior Error Probability (PEP)') \n",
    "ax.set_title('Peptide PEP Score distributions')\n",
    "## Remove top axes and right axes ticks\n",
    "ax.get_xaxis().tick_bottom()\n",
    "ax.get_yaxis().tick_left()\n",
    "fig.savefig('analysis/psm_pep_scores_cns.png', bbox_inches='tight')\n",
    "plt.show()\n",
    "fig.clf()\n",
    "\n",
    "pep_df = pd.DataFrame()\n",
    "for val in range(len(names)):\n",
    "    pep_df.loc[names[val], 'Count'] = len(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - median'] = np.median(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - std. dev.'] = np.std(pep_scores[val])\n",
    "\n",
    "pep_df.to_csv('analysis/cns_pep_scores.csv')\n",
    "pep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##post_hoc = list_kw_dunn(names, pep_scores, \"PEP\", \"Workflow\", 'analysis')\n",
    "#post_hoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_mapping = hmapping[hmapping['taxon_rank']=='species']\n",
    "primate_peptides = set(hmapping[hmapping['order_name']=='Primates']['peptide'].tolist())\n",
    "primate_non_human = primate_peptides & all_non_human_peps \n",
    "print(len(primate_non_human - contaminants))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2.5 Genome annotation refinement by orthologous peptide identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'human_results' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-21-d83f68207b2f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m \u001b[0;32min\u001b[0m \u001b[0morder\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0mpeps\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhuman_results\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'TargetPeptides'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0mpeps\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpeps\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpeps\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Potential contaminant'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;34m'+'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;31m#peps = peps[peps['PEP'] < 1]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'human_results' is not defined"
     ]
    }
   ],
   "source": [
    "pep_scores = []\n",
    "names = []\n",
    "\n",
    "for key in keys:\n",
    "    peps = human_results[key]['TargetPeptides']\n",
    "    peps = peps[peps['Potential contaminant'] != '+']\n",
    "    #peps = peps[peps['PEP'] < 1]\n",
    "    rpeps = human_results[key]['ReversePeptides']\n",
    "    key_peps = set(peps['Sequence'].tolist()).copy()\n",
    "    non_human= set(key_peps)\n",
    "    human = set(key_peps)\n",
    "    for qkey in keys:\n",
    "        if not qkey == key:\n",
    "            non_human = human & primate_non_human\n",
    "            human -= all_non_human_peps\n",
    "    print('{} Human: '.format(key), len(human))\n",
    "    print('{} Non-human primates: '.format(key), len(non_human))\n",
    "    human_pep = peps[peps['Sequence'].isin(human)]['PEP'].tolist()\n",
    "    pep_scores.append(human_pep)\n",
    "    names.append(key + ': Human')\n",
    "    non_human_pep = peps[peps['Sequence'].isin(non_human)]['PEP'].tolist()\n",
    "    pep_scores.append(non_human_pep)\n",
    "    names.append(key + ': Other primates')\n",
    "    reverse_pep = rpeps['PEP'].tolist()\n",
    "    pep_scores.append(reverse_pep)\n",
    "    names.append(key + ': Reverse')\n",
    "\n",
    "#############\n",
    "## BOXPLOT ##\n",
    "#############\n",
    "\n",
    "colours = ['b','g','r','c','m','y','k']\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "# Create the boxplot\n",
    "bp = ax.boxplot(pep_scores, patch_artist=True, showfliers=False)\n",
    "## change outline color, fill color and linewidth of the boxes\n",
    "count = 0\n",
    "col_ind=0\n",
    "for box in bp['boxes']:\n",
    "    count += 1\n",
    "    # change outline color\n",
    "    box.set( color='#7570b3', linewidth=1)\n",
    "    # change fill color\n",
    "    box.set( facecolor = colours[col_ind] )\n",
    "    ## change color and linewidth of the whiskers\n",
    "    if count % 3 == 0:\n",
    "        col_ind +=1\n",
    "for whisker in bp['whiskers']:\n",
    "    whisker.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the caps\n",
    "for cap in bp['caps']:\n",
    "    cap.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the medians\n",
    "for median in bp['medians']:\n",
    "    median.set(color='#b2df8a', linewidth=2)\n",
    "    #median.set(linewidth=2)\n",
    "## change the style of fliers and their fill\n",
    "for flier in bp['fliers']:\n",
    "    flier.set(marker='.', color='#e7298a', alpha=0.5)\n",
    "## Custom x-axis labels\n",
    "ax.set_xticklabels(names, rotation=90)\n",
    "#ax.set_yticklabels('Posterior Error Probability (PEP)') \n",
    "ax.set_title('Peptide PEP Score distributions')\n",
    "## Remove top axes and right axes ticks\n",
    "ax.get_xaxis().tick_bottom()\n",
    "ax.get_yaxis().tick_left()\n",
    "fig.savefig('analysis/psm_pep_scores_cns_primate.png', bbox_inches='tight')\n",
    "plt.show()\n",
    "fig.clf()\n",
    "\n",
    "pep_df = pd.DataFrame()\n",
    "for val in range(len(names)):\n",
    "    pep_df.loc[names[val], 'Count'] = len(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - median'] = np.median(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - std. dev.'] = np.std(pep_scores[val])\n",
    "\n",
    "pep_df.to_csv('analysis/cns_primate_pep_scores.csv')\n",
    "pep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(names)\n",
    "post_hoc = list_kw_dunn([names[0], \n",
    "                       \n",
    "                         names[2], \n",
    "                         names[3], \n",
    "                         names[4],\n",
    "                         names[5]], \n",
    "                        [pep_scores[0], \n",
    "                      \n",
    "                         pep_scores[2], \n",
    "                         pep_scores[3], \n",
    "                         pep_scores[4],\n",
    "                         pep_scores[5]], \"PEP\", \"Workflow\", 'analysis')\n",
    "post_hoc"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Query1                  AEGDISNVADGFK                  13\n",
    "                        AEGDISNVA+GFK\n",
    "Q9UBB4 ATX10_HUMAN359   AEGDISNVANGFK                  371\n",
    "\n",
    "Query1                  ASSSILINESEPTTNIQIR            19\n",
    "                        ASSSILINESEPTTNIQIR\n",
    "Q6ZQU3 Q6ZQU3_HUMAN56   ASSSILINESEPTTNIQIR            74\n",
    "\n",
    "Query1                  AVFVDLEPTVVEEVR                15\n",
    "                        AVFVDLEPTVV+EVR\n",
    "Q6PEY2 TBA3E_HUMAN65    AVFVDLEPTVVDEVR                79"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Adult stool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16  Raw file                                          FASP_A\n",
       "    Peptide Sequences Identified                        7450\n",
       "17  Raw file                                         FASP_AE\n",
       "    Peptide Sequences Identified                        9306\n",
       "18  Raw file                                          FASP_B\n",
       "    Peptide Sequences Identified                        5058\n",
       "19  Raw file                                         FASP_BE\n",
       "    Peptide Sequences Identified                        9471\n",
       "20  Raw file                                          FASP_C\n",
       "    Peptide Sequences Identified                        4213\n",
       "21  Raw file                                         FASP_CE\n",
       "    Peptide Sequences Identified                        7128\n",
       "22  Raw file                                          FASP_D\n",
       "    Peptide Sequences Identified                        4613\n",
       "23  Raw file                                         FASP_DE\n",
       "    Peptide Sequences Identified                        8829\n",
       "24  Raw file                                         Insol_A\n",
       "    Peptide Sequences Identified                        8094\n",
       "25  Raw file                                        Insol_AE\n",
       "    Peptide Sequences Identified                        7988\n",
       "26  Raw file                                         Insol_B\n",
       "    Peptide Sequences Identified                        6115\n",
       "27  Raw file                                        Insol_BE\n",
       "    Peptide Sequences Identified                        8236\n",
       "28  Raw file                                         Insol_C\n",
       "    Peptide Sequences Identified                        7984\n",
       "29  Raw file                                        Insol_CE\n",
       "    Peptide Sequences Identified                        9051\n",
       "30  Raw file                                         Insol_D\n",
       "    Peptide Sequences Identified                        7478\n",
       "31  Raw file                                        Insol_DE\n",
       "    Peptide Sequences Identified                       10093\n",
       "32  Raw file                                           Total\n",
       "    MS                                                128564\n",
       "    MS/MS                                             815871\n",
       "    MS3                                                    0\n",
       "    MS/MS Submitted                                   961839\n",
       "    MS/MS Submitted (SIL)                             669903\n",
       "    MS/MS Submitted (ISO)                                  0\n",
       "    MS/MS Submitted (PEAK)                            291936\n",
       "    MS/MS Identified                                  182635\n",
       "    MS/MS Identified (SIL)                            176251\n",
       "    MS/MS Identified (ISO)                                 0\n",
       "    MS/MS Identified (PEAK)                             6384\n",
       "    MS/MS Identified [%]                               18.99\n",
       "    MS/MS Identified (SIL) [%]                         26.31\n",
       "    MS/MS Identified (PEAK) [%]                         2.19\n",
       "    Peptide Sequences Identified                       29943\n",
       "    Peaks                                        1.64477e+07\n",
       "    Isotope Patterns                              1.7284e+06\n",
       "    Isotope Patterns Sequenced                        496726\n",
       "    Isotope Patterns Sequenced (z>1)                  475577\n",
       "    Isotope Patterns Sequenced [%]                     28.74\n",
       "    Isotope Patterns Sequenced (z>1) [%]               32.62\n",
       "    Isotope Patterns Repeatedly Sequenced             121529\n",
       "    Isotope Patterns Repeatedly Sequenced [%]          24.47\n",
       "    Av. Absolute Mass Deviation [ppm]                0.77932\n",
       "    Mass Standard Deviation [ppm]                     1.0776\n",
       "    Av. Absolute Mass Deviation [mDa]                0.57322\n",
       "    Mass Standard Deviation [mDa]                    0.84464\n",
       "dtype: object"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# MetaNovp\n",
    "summary = pd.read_csv('metanovo_stool/stool_analysis_txt/txt/summary.txt',sep='\\t')\n",
    "summary[16:].stack()\n",
    "#parameters = pd.read_csv('txt_metanovo/parameters.txt',sep='\\t')\n",
    "#pramaters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "{'Afrotheria',\n",
    " 'Atherinomorphae',\n",
    " 'Batrachia',\n",
    " 'Characiphysae',\n",
    " 'Cypriniphysae',\n",
    " 'Euarchontoglires',\n",
    " 'Laurasiatheria',\n",
    " 'Lepidosauria',\n",
    " 'Neognathae',\n",
    " 'Octopodiformes',\n",
    " 'Palaeognathae',\n",
    " 'Parasitiformes',\n",
    " 'Podoplea',\n",
    " nan}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Analysis: metanovo - UniProt\n",
      "metanovo - UniProt: Total peptides:  29943\n",
      "metanovo - UniProt: Total target peptides:  29874\n",
      "metanovo - UniProt: proteins:  6452\n",
      "metanovo - UniProt: Total target proteins:  6385\n",
      "\n"
     ]
    }
   ],
   "source": [
    "folders = {'metanovo - UniProt':'metanovo_stool/stool_analysis_txt/txt'}\n",
    "\n",
    "stool_results = {}\n",
    "order = list(folders.keys())\n",
    "for name in folders:\n",
    "    path = folders[name]\n",
    "    res = process_txt(path, name)\n",
    "    stool_results[name] = res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     253\r\n"
     ]
    }
   ],
   "source": [
    "cat metanovo_stool/stool_analysis_txt/txt/analysis/peptides/contaminant_peptides_list.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   29873\r\n"
     ]
    }
   ],
   "source": [
    "cat metanovo_stool/stool_analysis_txt/txt/analysis/peptides/target_peptides_list.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "contaminant_peptides_list.txt    reverse_peptides_list.txt\r\n",
      "non_reference_peptides_list.txt  target_peptides.txt\r\n",
      "reference_peptides_list.txt      target_peptides_list.txt\r\n"
     ]
    }
   ],
   "source": [
    "ls metanovo_stool/stool_analysis_txt/txt/analysis/peptides"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    2558\r\n"
     ]
    }
   ],
   "source": [
    "cat metanovo_stool/stool_analysis_txt/txt/analysis/peptides/reference_peptides_list.txt | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metanovo - UniProt\n"
     ]
    }
   ],
   "source": [
    "##################\n",
    "## Peptide Sets ##\n",
    "##################\n",
    "\n",
    "all_human_peps = set()\n",
    "all_non_human_peps = set()\n",
    "\n",
    "keys = list(stool_results.keys())\n",
    "names = []\n",
    "pep_scores = []\n",
    "for key in keys:\n",
    "    print(key)\n",
    "    peps = set(stool_results[key]['TargetPeptides']['Sequence'])\n",
    "    for pep in peps:\n",
    "        if pep in seqs:\n",
    "            all_human_peps.add(pep)\n",
    "        else:\n",
    "            all_non_human_peps.add(pep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "metanovo - UniProt Human:  2559\n",
      "metanovo - UniProt Non-human:  27061\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAGmCAYAAABiEMNdAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3Xm8XHV9//HXm4RgyioQLXtQgjZYRA2gFVsrVcHKogKGVkVF8deKLK1SUBG0okKt2Kq1RdmKCyCCBoqlCLhVBRKIImBqZBGCS9gX2QLv3x/ne2EyzD13cjNzz5257+fjMY/M+Z5lPuckmc98l/M9sk1ERMRo1mg6gIiImNySKCIiolYSRURE1EqiiIiIWkkUERFRK4kiIiJqJVFE30i6X9KzRln3Vkk/mOiYhk37day75uM49vslfbG8ny3Jkqb36Nhbllin9eJ40V9JFFOMpJskPVj+k/5W0mmS1unBcb8j6R2tZbbXsX3D6h57lM96qJzD7ZLOlbRJWXeapEfKupHXT8q6kS+7kfKbJB1Z8zkHSvq5pPvKtbpQ0rq9Pp9e6uaaS3q5pFu7ONbHbL9jrO26Ua71X7Qc+1cl1sd6cfzorySKqWkP2+sALwTmAR9sOJ7xOLicw7bABsCJLetOKF9CI6/nt+27Qdl3f+BDknZrP7ikPwM+Buxve13gj4CzenkCvfp13g+TObaYeEkUU5jtZcC3gOcBSFpf0smSfi1pmaSPjjQNlCaO/5X0WUn3lF/au5Z1xwEvAz5bfql/tpRb0jbl/UaSFki6V9IVwLNbY5H0XEkXS7pT0hJJ+3V5DncCXx85h1U8/x8B146y747Aj2xfPfI5tk+3fV+Jd6akf5Z0c7keP5A0s6zbU9K1ku4utZ8/ajnPmyT9g6SfAg9Imi5pU0lfl7Rc0o2SDhkt5i6uY+s1f42k60qNaJmk90pam+rvfNOWmtWmko6VdI6kL0m6F3hrKftSWwhvl3Rb+Tfy3pbPPU3SR1uWn6i1SDoD2BI4v3zeEe1NWSWGBeXvf6mkd7Yc61hJZ0v6z3Iu10qa17L+H8r53Vf+7ew62vWL8UmimMIkbQG8Bri6FJ0GrAC2AV4AvApobXrYGfglsDFwDHCupA1tfwD4PuVXvu2DO3zc54CHgE2At5fXSBxrAxcDXwGeAcwH/k3S3C7OYWPgDS3n0BVVXgpsN8q+lwOvlvRhSS+VtFbb+k8CLwL+BNgQOAJ4XNK2wFeBw4BZwIVUX5AzWvbdH/hLqprQ48D5wE+AzYBdgcMkvXqU0Ee9jh2cDLyr1IieB1xq+wFgd+C2lhrXbWX7vYBzSlxfHuWYfw7Mofq38Q+tzUmjsf1m4FeUmqztEzpsdiZwK7ApsA/wMUmvaFm/Z9lmA2ABMPJj5DnAwcCO5TxfDdw0VkyximznNYVeVP+J7gfuBm4G/g2YCTwTeBiY2bLt/sBl5f1bgdsAtay/Anhzef8d4B1tn2WqpDMNeBR4bsu6jwE/KO/fCHy/bd//AI4Z5Ry+A/y+nMMyqi+1WWXdaVRfpHe3vE4v62aXmO4G7gKuBw6puVa7U32J312u2afKuawBPAg8v8M+RwNntyyvUWJ8ecv1f3vL+p2BX7Ud4yjg1A7Hrr2Orde8vP8V8C5gvbbjvBy4ta3sWOB7Hcq+1HbtWj/7BODkluv+0dE+o5z3X7QsjxxvOrAF8Biwbsv6jwOntcTx7ZZ1c4EHy/ttgN8BfwGs2fT/r2F9pR1yatrb9rdbCyT9MbAm8GtJI8VrALe0bLbM5X9ncTPVL8CxzKL6Qmg91s0t77cCdpZ0d0vZdOCMmmMeYvuLo6z7pO26fpeNba+oCxjA9reAb0lag+qX9NeAJcB5wNOoalftNqXl3Gw/LukWqtrCiNbrsBVVM1DruU+jqqG1G+s6tnsDVf/TJ0pT15GumttGc0vNuk7b3Az8cRf7jGVT4E6XZr2WY89rWf5Ny/vfA0+TNN32UkmHUSWT7SRdBPydn6wlRQ+k6SlG3EJVo9jY9gbltZ7t7Vq22UwtWYSq3XnkP2TdNMTLqZq0tmjbt/Wzv9vyuRu4aqL4m/GfTu/Yftz2JcClVE04t1PVWp7dYfPbqL78gaqJi+q8l7UesuX9LcCNbee+ru3XdDj2WNexPe4rbe9F1Zz3DeDsDp+/0i6jHatF+2eP/P0/APxBy7o/XIVj3wZsqJVHlG3JytdsVLa/YnsXqutu4Phu9ovuJVEEALZ/DfwP8M+S1pO0hqRnqxr9M+IZwCGS1pS0L9VIoAvLut8CHcfvuxoCeS5wrKQ/KH0PB7RscgGwraQ3l2OvKWnH1k7giSZpL0nzJT299GfsBPwZ8GPbjwOnAJ8qnbDTJL2k9GOcDfylpF0lrQn8PVUC/uEoH3UFcF/pkJ1ZjvU8STu2b9jFdWyNf4akv5a0vu1HgXup+kOg+rvaSNL647g0R5fP3g54G0+OBFsMvEbShpL+kKqPplXdv49bqK7PxyU9TdL2wIFAe0f6U0h6jqRXlGv/EFWT4ONj7BarKIkiWr0FmAFcR9WGfw5Vp+mIy6k6Mm8HjgP2sX1HWfcvwD6S7pL0rx2OfTCwDlUTwmnAqSMrSpPDq6g6sW8r2xwPtHcgd+sIrXwfxe3jOMZdwDuBX1B9yX4J+CfbI5287wWuAa4E7izxrmF7CfAm4DNU12kPqk7cRzp9SPnyfy2wA3Bj2eeLwGhf4qNexw7eDNxURjH9P+Cvy2f+nKrD/QZVI7O6aT4c8V1gKXAJVRPf/5TyM6g65G+i+sHRPpT448AHy+e9l6fan6rf4jaqpr1j2ptHR7EW8Amq6/Ybqh8zR63C+UQXtHKTc0Rnkt5K1Vm9S9OxRMTESo0iIiJqJVFEREStND1FREStrmoUknYrt8YvVYdJ1CStJemssv5ySbNL+UaSLlPLtA6lfF1Ji1tet0v6dFn3VlVTGYys68mkZBERMT5j3nCnaq6fzwGvpLrF/kpJC2xf17LZgcBdtreRNJ9qBMgbqYarHU019vyJ+XTKKJcdWj5jEdWwvxFnufM0EBERMcG6uTN7J2Cpy9TFks6kmhOmNVHsRXVnJFRDKj8rSa7mlfmByiRlnZS5cZ5B5ztRu7Lxxht79uzZ4909ImJKWrRo0e22Z421XTeJYjNWvm3/Vqr5aTpuY3uFpHuAjajGNo9lPlUNorWz5A2S/hT4P+DwckPOqGbPns3ChQu7+KiIiBghqW4KmCdMhlFP86lu/hlxPjDb9vZUM4qe3mknSQdJWihp4fLlyycgzIiIqambRLGMled32ZynzsHyxDaq5pdfH7iDMUh6PjDd9qKRMtt32H64LH6Rairnp7B9ku15tufNmjVmzSkiIsapm0RxJTBH0tZlTv35VPPBt1rAk3PO7EM173034273Z+XaBCqPtCz2pJoKOiIiGjJmH0XpczgYuIhq+uNTbF8r6SPAQtsLqB6QcoakpVTz3swf2V/STcB6wAxJewOvahkxtR/Vg3NaHSJpT6pZMu+keg5CREQ0ZChuuJs3b57TmR0RsWokLbI9b6ztJkNndkRETGJJFBERUSuJIiIiauWZ2THpHHH4BT093gknvranx4uYapIoYtLp9ov9iMMvSBKImABpeoqIiFpJFBERUSuJIiIiaiVRRERErSSKiIiolUQRERG1kigiIqJWEkVERNRKooiIiFpJFBERUSuJIiIiaiVRRERErSSKiIiolUQRERG1kigiIqJWV4lC0m6SlkhaKunIDuvXknRWWX+5pNmlfCNJl0m6X9Jn2/b5Tjnm4vJ6Rt2xIiKiGWMmCknTgM8BuwNzgf0lzW3b7EDgLtvbACcCx5fyh4CjgfeOcvi/tr1Def1ujGNFREQDuqlR7AQstX2D7UeAM4G92rbZCzi9vD8H2FWSbD9g+wdUCaNbHY+1CvtHREQPdZMoNgNuaVm+tZR13Mb2CuAeYKMujn1qaXY6uiUZjPdYERHRB012Zv+17T8GXlZeb16VnSUdJGmhpIXLly/vS4AREdFdolgGbNGyvHkp67iNpOnA+sAddQe1vaz8eR/wFaomrq6PZfsk2/Nsz5s1a1YXpxEREePRTaK4EpgjaWtJM4D5wIK2bRYAB5T3+wCX2vZoB5Q0XdLG5f2awGuBn43nWBER0V/Tx9rA9gpJBwMXAdOAU2xfK+kjwELbC4CTgTMkLQXupEomAEi6CVgPmCFpb+BVwM3ARSVJTAO+DXyh7DLqsSIiYuKNmSgAbF8IXNhW9qGW9w8B+46y7+xRDvuiUbYf9VgRETHxcmd2RETUSqKIiIhaSRQREVEriSIiImolUURERK0kioiIqJVEERERtZIoIiKiVhJFRETUSqKIiIhaXU3hERGD64jDL+jp8U448bU9PV5MfkkUEUOumy/2Iw6/IAkgRpWmp4iIqJVEERERtZIoIiKiVhJFRETUSqKIiIhaSRQREVEriSIiImolUURERK0kioiIqNVVopC0m6QlkpZKOrLD+rUknVXWXy5pdinfSNJlku6X9NmW7f9A0n9J+rmkayV9omXdWyUtl7S4vN6x+qcZERHjNWaikDQN+BywOzAX2F/S3LbNDgTusr0NcCJwfCl/CDgaeG+HQ3/S9nOBFwAvlbR7y7qzbO9QXl9cpTOKiIie6qZGsROw1PYNth8BzgT2attmL+D08v4cYFdJsv2A7R9QJYwn2P697cvK+0eAq4DNV+M8IiKiT7pJFJsBt7Qs31rKOm5jewVwD7BRNwFI2gDYA7ikpfgNkn4q6RxJW3RznIiI6I9GO7MlTQe+Cvyr7RtK8fnAbNvbAxfzZE2lfd+DJC2UtHD58uUTE3BExBTUTaJYBrT+qt+8lHXcpnz5rw/c0cWxTwJ+YfvTIwW277D9cFn8IvCiTjvaPsn2PNvzZs2a1cVHRUTEeHSTKK4E5kjaWtIMYD6woG2bBcAB5f0+wKW2XXdQSR+lSiiHtZVv0rK4J3B9FzFGRESfjPngItsrJB0MXARMA06xfa2kjwALbS8ATgbOkLQUuJMqmQAg6SZgPWCGpL2BVwH3Ah8Afg5cJQngs2WE0yGS9gRWlGO9tUfnGhER49DVE+5sXwhc2Fb2oZb3DwH7jrLv7FEOq1G2Pwo4qpu4IiKi/3JndkRE1EqiiIiIWkkUERFRK4kiIiJqJVFEREStJIqIiKiVRBEREbWSKCIiolYSRURE1EqiiIiIWkkUERFRK4kiIiJqJVFEREStJIqIiKiVRBEREbWSKCIiolYSRURE1EqiiIiIWkkUERFRK4kiIiJqdZUoJO0maYmkpZKO7LB+LUlnlfWXS5pdyjeSdJmk+yV9tm2fF0m6puzzr5JUyjeUdLGkX5Q/n776pxkREeM1ZqKQNA34HLA7MBfYX9Lcts0OBO6yvQ1wInB8KX8IOBp4b4dDfx54JzCnvHYr5UcCl9ieA1xSliMioiHd1Ch2ApbavsH2I8CZwF5t2+wFnF7enwPsKkm2H7D9A6qE8QRJmwDr2f6xbQP/Cezd4Vint5RHREQDukkUmwG3tCzfWso6bmN7BXAPsNEYx7x1lGM+0/avy/vfAM/sIsaIiOiTSd2ZXWob7rRO0kGSFkpauHz58gmOLCJi6ugmUSwDtmhZ3ryUddxG0nRgfeCOMY65+SjH/G1pmhppovpdpwPYPsn2PNvzZs2a1cVpRETEeHSTKK4E5kjaWtIMYD6woG2bBcAB5f0+wKWlNtBRaVq6V9KLy2intwDf7HCsA1rKIyKiAdPH2sD2CkkHAxcB04BTbF8r6SPAQtsLgJOBMyQtBe6kSiYASLoJWA+YIWlv4FW2rwP+FjgNmAl8q7wAPgGcLelA4GZgv16caEREjM+YiQLA9oXAhW1lH2p5/xCw7yj7zh6lfCHwvA7ldwC7dhNXRET0X1eJIsZ2xOEX9OxYJ5z42p4dKyJidSVR9Eg3X+5HHH5BkkBEDJxJPTw2IiKal0QRERG1kigiIqJWEkVERNRKooiIiFpJFBERUSuJIiIiaiVRRERErSSKiIiolUQRERG1kigiIqJWEkVERNRKooiIiFpJFBERUSuJIiIiauV5FBERq2AqPqQsiSIiYhVMxYeUpekpIiJqJVFEREStrhKFpN0kLZG0VNKRHdavJemssv5ySbNb1h1VypdIenUpe46kxS2veyUdVtYdK2lZy7rX9OZUIyJiPMbso5A0Dfgc8ErgVuBKSQtsX9ey2YHAXba3kTQfOB54o6S5wHxgO2BT4NuStrW9BNih5fjLgPNajnei7U+u/ulFRMTq6qZGsROw1PYNth8BzgT2attmL+D08v4cYFdJKuVn2n7Y9o3A0nK8VrsCv7R983hPIiIi+qebRLEZcEvL8q2lrOM2tlcA9wAbdbnvfOCrbWUHS/qppFMkPb1TUJIOkrRQ0sLly5d3cRoRETEejXZmS5oB7Al8raX488CzqZqmfg38c6d9bZ9ke57tebNmzep7rBERU1U3iWIZsEXL8ualrOM2kqYD6wN3dLHv7sBVtn87UmD7t7Yfs/048AWe2lQVERETqJtEcSUwR9LWpQYwH1jQts0C4IDyfh/gUtsu5fPLqKitgTnAFS377U9bs5OkTVoWXwf8rNuTiYiI3htz1JPtFZIOBi4CpgGn2L5W0keAhbYXACcDZ0haCtxJlUwo250NXAesAN5t+zEASWtTjaR6V9tHniBpB8DATR3WR0TEBOpqCg/bFwIXtpV9qOX9Q8C+o+x7HHBch/IHqDq828vf3E1MERExMXJndkRE1EqiiIiIWkkUERFRK9OMx4Q65v2X8OCDD/bseL14NsDMmTP58Md27UE0EcMpiSIm1IMPPsg/fXpyzdP/vsN69yCaiGGUpqeIiKiVGkVEBJOzWRQmR9NoEkVEBJOzWRQmR9Nomp4iIqJWEkVERNRKooiIiFpJFBERUSuJIiIiaiVRRERErSSKiIiolfsoIgZUbhCLiZJEETGgcoNYTJQ0PUVERK0kioiIqJVEERERtbpKFJJ2k7RE0lJJR3ZYv5aks8r6yyXNbll3VClfIunVLeU3SbpG0mJJC1vKN5R0saRflD+fvnqnGBERq2PMRCFpGvA5YHdgLrC/pLltmx0I3GV7G+BE4Piy71xgPrAdsBvwb+V4I/7c9g6257WUHQlcYnsOcElZjoiIhnRTo9gJWGr7BtuPAGcCe7Vtsxdwenl/DrCrJJXyM20/bPtGYGk5Xp3WY50O7N1FjBER0SfdJIrNgFtalm8tZR23sb0CuAfYaIx9DfyPpEWSDmrZ5pm2f13e/wZ4ZhcxRkREnzR5H8UutpdJegZwsaSf2/5e6wa2Lcmddi7J5SCALbfcsv/RRkRMUd3UKJYBW7Qsb17KOm4jaTqwPnBH3b62R/78HXAeTzZJ/VbSJuVYmwC/6xSU7ZNsz7M9b9asWV2cRkREjEc3ieJKYI6krSXNoOqcXtC2zQLggPJ+H+BS2y7l88uoqK2BOcAVktaWtC6ApLWBVwE/63CsA4Bvju/UIiKiF8ZserK9QtLBwEXANOAU29dK+giw0PYC4GTgDElLgTupkgllu7OB64AVwLttPybpmcB5VX8304Gv2P7v8pGfAM6WdCBwM7BfD883IuIpzr/pn3jZofCyQ69vOpRRNDtVS1d9FLYvBC5sK/tQy/uHgH1H2fc44Li2shuA54+y/R1AZhSLiJgkMilgREx5e8x+H0ccfsGknWRxjxObjSFTeERERK0kioiIqJVEERERtZIoIiKiVhJFRETUSqKIiIhaSRQREVEriSIiImrlhrsxHPP+S3jwwQd7drwjDr+gJ8eZOXMmH/5YbmCPiP5LohjDgw8+OGnv1oyImAhpeoqIiFpJFBERUSuJIiIiaiVRRERErXRmR0RQjSScjINEZs6c2XQISRQREUBPh5sfcfgFnHDi5BstOV5peoqIiFpJFBERUSuJIiIianWVKCTtJmmJpKWSjuywfi1JZ5X1l0ua3bLuqFK+RNKrS9kWki6TdJ2kayUd2rL9sZKWSVpcXq9Z/dOMiIjxGrMzW9I04HPAK4FbgSslLbB9XctmBwJ32d5G0nzgeOCNkuYC84HtgE2Bb0vaFlgB/L3tqyStCyySdHHLMU+0/clenWRERIxfNzWKnYCltm+w/QhwJrBX2zZ7AaeX9+cAu0pSKT/T9sO2bwSWAjvZ/rXtqwBs3wdcD2y2+qcTERG91k2i2Ay4pWX5Vp76pf7ENrZXAPcAG3Wzb2mmegFweUvxwZJ+KukUSU/vIsaIiOiTRjuzJa0DfB04zPa9pfjzwLOBHYBfA/88yr4HSVooaeHy5csnJN6IiKmom0SxDNiiZXnzUtZxG0nTgfWBO+r2lbQmVZL4su1zRzaw/Vvbj9l+HPgCVdPXU9g+yfY82/NmzZrVxWlERMR4dJMorgTmSNpa0gyqzukFbdssAA4o7/cBLrXtUj6/jIraGpgDXFH6L04Grrf9qdYDSdqkZfF1wM9W9aQiIqJ3xhz1ZHuFpIOBi4BpwCm2r5X0EWCh7QVUX/pnSFoK3EmVTCjbnQ1cRzXS6d22H5O0C/Bm4BpJi8tHvd/2hcAJknYADNwEvKuH5xsREauoq7meyhf4hW1lH2p5/xCw7yj7Hgcc11b2A0CjbP/mbmKKiIiJkTuzIyKiVhJFRETUSqKIiIhaSRQREVEriSIiImolUURERK0kioiIqJVEERERtZIoIiKiVhJFRETUSqKIiIhaSRQREVGrq0kBI3rh/Jv+iZcdCi879PqmQ+ngtU0HsEom97WEQbueUS81ioiIqJUaRUyYPWa/jyMOv4B/+vTk+rX5vsMuYI8Tm45i1UzWawmDeT2jXhJFRMQqOOLwC3q23QknTr5E30kSRUTEKhiUL/deSh9FRETUSqKIiIhaSRQREVEriSIiImp1lSgk7SZpiaSlko7ssH4tSWeV9ZdLmt2y7qhSvkTSq8c6pqStyzGWlmPOWL1TjIiI1THmqCdJ04DPAa8EbgWulLTA9nUtmx0I3GV7G0nzgeOBN0qaC8wHtgM2Bb4taduyz2jHPB440faZkv69HPvzvTjZVZW7XyMiuqtR7AQstX2D7UeAM4G92rbZCzi9vD8H2FWSSvmZth+2fSOwtByv4zHLPq8ox6Acc+/xn15ERKyubu6j2Ay4pWX5VmDn0baxvULSPcBGpfzHbftuVt53OuZGwN22V3TYfsLl7teYzGbOnMn7Duvu5q+JNHPmzKZDiB4b2BvuJB0EHASw5ZZb9u1z8p+xtybj9RzUa/nhj+3a1Xbd3kncral4w9lU102iWAZs0bK8eSnrtM2tkqYD6wN3jLFvp/I7gA0kTS+1ik6fBYDtk4CTAObNm+cuzmNcuv3PGN3J9Zx4+WKP1dVNH8WVwJwyGmkGVef0grZtFgAHlPf7AJfadimfX0ZFbQ3MAa4Y7Zhln8vKMSjH/Ob4Ty8iIlbXmDWK0udwMHARMA04xfa1kj4CLLS9ADgZOEPSUuBOqi9+ynZnA9cBK4B3234MoNMxy0f+A3CmpI8CV5djR0REQ1T9iB9s8+bN88KFC5sOIyJioEhaZHveWNvlzuyIiKiVRBEREbWSKCIiolYSRURE1BqKzmxJy4Gbm46jCxsDtzcdxBDJ9eydXMveGpTruZXtWWNtNBSJYlBIWtjNCIPoTq5n7+Ra9tawXc80PUVERK0kioiIqJVEMbFOajqAIZPr2Tu5lr01VNczfRQREVErNYqIiKiVRBEREbWSKGIgqLLF2FtGN8q0/2OWRUASRQyI8qySC5uOY4h8vUPZOR3KoguSninpZEnfKstzJR3YdFy9kkTRR5JeL+kXku6RdK+k+yTd23RcA+wqSTs2HcQgk/RcSW8A1i//PkdebwWe1nB4g+w0qufrbFqW/w84rLFoemxgn5k9IE4A9rB9fdOBDImdgb+WdDPwACCqysb2zYY1UJ4DvBbYANijpfw+4J2NRDQcNrZ9tqSj4IkHvj3WdFC9kkTRX79NkuipVzcdwKCz/U3gm5JeYvtHTcczRB6QtBFgAEkvBu5pNqTeyX0UfSTpX4A/BL4BPDxSbvvcxoIacJJ2AebYPlXSLGAd2zc2HdegKdfuncBsWn4w2n57UzENMkkvBD4DPA/4GTAL2Mf2TxsNrEeSKPpI0qkdip3/jOMj6RhgHvAc29tK2hT4mu2XNhzawJH0Q+D7wCLgiSYS2506uaMLkqZTNe0JWGL70YZD6pkkihgYkhYDLwCusv2CUvbT9FGsOkmLbe/QdBzDQtK+wH/bvk/SB4EXAh+1fVXDofVE+ij6SNLTgAOB7WgZUZIaxbg9YtuSRtqB1246oAF2gaTX2M6Q49442vbXStPorsAngc9TDcAYeBke219nUPVRvBr4LrA51eiSGJ+zJf0HsIGkdwLfBr7QcEyD6lCqZPFghm73xEjz3V8CX7D9X8CMBuPpqTQ99ZGkq22/YKR5RNKawPdtv7jp2AaVpFcCr6JqB77I9sUNhxSBpAuAZcArqZqdHgSusP38RgPrkTQ99ddIZ9bdkp4H/AZ4RoPxDLySGJIcekDS04E5rNws+r3mIhpo+wG7AZ+0fbekTYD3NRxTzyRR9NdJ5T/j0cACYB3gQ82GNLgkvR44nirZiidvuFuv0cAGkKR3UDU/bQ4sBl4M/Ah4RZNxDSJJ06gGWDx3pMz2r4FfNxdVb6XpKQaGpKXkTveekHQNsCPwY9s7SHou8DHbr284tIEk6ZvAe2z/qulY+iE1ij6StAHwFp56U9MhTcU04HKne+88ZPshSUhay/bPJT2n6aAG2NOBayVdQTW9DAC292wupN5JouivC4EfA9cAjzccy8AqTU4ACyWdRe5074Vbyw+ZbwAXS7oLuLnhmAbZ0U0H0E9peuojSVfZfmHTcQy6Ue5wH5E73VeTpD8D1qe6YeyRpuMZVJK2oppe5tuS/gCYZnsohsMnUfSRpMOB+4ELWPkX8J2NBTXAJL3U9v+OVRbdKZ2wz2TlZtGhbGPvt3Jfz0HAhrafLWkO8O+2d204tJ5IougjSe8GjgPupswqSfUL+FnNRTW4OtXQUmsbH0nvAY4BfsuTzaKZsn2cyvQi6SmaAAAU10lEQVQyOwGXt0wvc43tP242st5IH0V//T2wje3bmw5kkEl6CfAnwCxJf9eyaj1gWjNRDbxDqSZXvKPpQIbEw7YfkQQ8MUHg0PwKT6Lor6XA75sOYgjMoLoHZTqwbkv5vcA+jUQ0+G5hiJ6XMAl8V9L7gZll9oC/Bc5vOKaeSdNTH0k6j2pCwMtYuY8iw2PHQdKWwJ0Atu9vOJyBJulkqimx/4uV/21+qrGgBpikNagmAH1iehngix6SL9jUKPrrG+UVq0nS3wBHAWuX5fuB423/W6OBDa5fldcMhmjyugbtDfyn7aGcpDI1ipj0yvz+fwIcbPuGUvYs4F+oOg8/2mR8EWUI9yuA7wFnUQ01XtFsVL2TRNFHkm6kQ4dWRj2tGklLgOfbfqitfCbwE9vbNhPZ4JH0aduHSTqfzv82h+JO4iaU2aF3B94I7AJcbPsdzUbVG2l66q95Le+fBuwLbNhQLIPM7UmiFD4oKXe8r5ozyp+fbDSKIWT7UUnfokrAM6mao4YiUaRGMcEkLbL9oqbjGCSSLqGasO6StvJXUD1Z7M+biSyiImmkJvFy4DvA2cD/DEvzU2oUfSSp9UawNahqGLnmq+4Q4JuSfgAsKmXzgJcCezUW1QCT9FrgH4GtqP5NZsr21fMWqr6Jd9l+eKyNB01qFH0k6bKWxRXATVQPNlnSTESDqzx//K+ohhsDXAd8uVOTVIytTNn+euCaYRnC2bS2uZ5mAtMz11NEDKzyI2ZX2+nj6YFhn+spzSB90DbNxFPkpqaYBI4ALpT0XXLDXS+8mzLXE4DtX0gamsceJ1H0R+s0E+8C/qOpQCJGcRzVzMZPIzfc9cJQz/WUpqc+k3T1yGySEZOFpJ/Zfl7TcQwLSSdQzRL9FuA9VHM9XWf7A40G1iNrNB3AFJBM3EOSjq1bjq5dKOlVTQcxRI4EllM9zfJdVE+3/GCjEfVQahR9lucl9JakPWyfP9pydEfSfVTzZj0MPEqGx/bcMD1UK30UfSDpGp6sSWwj6acjq8jDYcat/MdrTwp5WuA42F537K1iLOUpgfsBm1HN7/Szco/K+6nuzh6KZufUKPqgjKcele08xH4c8oS73pL0dGAOVYc2ALa/11xEg0fSacAWwBXAzsBtVDeDHml7aGaOTo2iD5IIeitPuOs9Se+gesrd5sBi4MXAj6hmQI3uzQO2t/14uSn0N8Czh+3JgenMjkHQ/oS7kVeecDd+hwI7AjeXubJeQDVqJ1bNIyM3LZZZAm4YtiQBaXqKASJpK9s3S1oH8pS71SHpSts7SloM7Gz7YUnX2t5uzJ3jCZJ+T/XIY6j6IJ9dloeqPzJNTzFI1pV0NWWqdkm3AwfY/lmzYQ2kWyVtQPUExosl3QWkyXTV/VHTAUyE1Cj6TNKxto8dbTm6J+mHwAdsX1aWX041/fifNBrYgJP0Z8D6wLdsP9p0PDH5pI+i/xaNsRzdW3skSQDY/g7lGdqxaiSNPMAI29+1vQA4pcGQYhJL01MfZdx/z90g6WiefErbm4AbGoxnkK3UF1HuB8gDtaKj1Cj66zNdlkV33g7MAs4Fvg5sXMqiS5KOKndlby/p3vK6D/gd8M2Gw4tJKn0UfdAy7v8w4MSWVesBr7P9/EYCG2DlF+/xtt/bdCzDQNLHbR/VdBzDZJj7I9P01B/t4/5HZNz/ONl+TNIuTccxLGwfJWkznnwU6kh57swev6Htj0yNoo8y7r+3JH2eak6drwEPjJTbPrexoAaUpE8A86keKftYKbbtPZuLanB1mgBwmCYFTKLoI0nPo+p43bAUZdz/apB0aodi204/xSqStIRq6omHx9w4xjTs85Cl6am/TgL+rm3c/0lU/Rex6t5n+/amgxgSNwBr0vIY1Fh1U2UesiSK/nrKuH9JGfe/iiTtQTXG/1FJjwP72f5hw2ENut8DiyVdwsrPzD6kuZAG0pToj0zTUx9JOg+4ipXH/b/I9uuai2rwlOd57Gf755J2Bk6w/WdNxzXIJB3Qqdz26RMdyzAY9v7I1Cj66+3Ah6nG/Rv4Phn3Px4rbP8cwPblkvLQndVk+3RJM4BtS9GSTN+xWoZ6HrIkij4p4/4/kKp8Tzyjrf13pWXbn2ogpoFW+stOB26imul0C0kHZHjsuA11f2QSRZ9k3H9PfYGV23/bl2PV/TPwKttLACRtC3yVTOMxXkPdH5lE0V9XS1pAxv2vFtsfbjqGIbTmSJIAsP1/ktZsMqABN9TzkKUzu48y7r83JB1h+wRJn6Hq61lJmvdWnaRTgMeBL5WiNwFr5N/m+JTnj38Y2IUn+yM/bPuuRgPrkdQo+ivj/nvj+vLnwkajGC5/A7wbGEmy3wM+31w4g2sq9EemRtEHreP+qX61Zdx/TAqSZgGzbF/XVr4d8Dvby5uJbLBJ+rHtFzcdR7+kRtEfxwEvax33D2Tc/2oqHa7vBWaz8kR2r2gqpgH0GeDfOpRvCHwA+KuJDWdoDHV/ZBJFf2Tcf398Dfh34Is8OZFdrJptOg2Btf39MulijM/TgDuA1h8tprqHauAlUfRHxv33xwrb+TJbPXU/WjLqafyGuj8yT7jrj5Fx/iOv9uUYn/Ml/a2kTSRtOPJqOqgBs1TSa9oLJe3OEA3nnCiS9pC0HPippFslDcUNdu3SmR0DQ9KNHYpt+1kTHsyAkjQH+C/ghzz5YJ15wEuA19r+v6ZiG0RTZR6yJIo+yLj/mMwkrUXVaf28UnQt8BXbDzUX1WBqf+bEMD2DolX6KPoj4/57SNLr24pM9RCoxbbvayCkgVYeVtTpZtBYdVOiPzI1ipj0RrnDfUNge+BA25dOcEgRAEg6pm79sEw/k0TRRxn331+StgLOtr1z07FEDLM0PfVXxv33UXlQTIZ0RmOmSn9kEkV/Zdx/H0l6Dnnm87hJOtb2saMtR1emRH9kEkV/nS/pb4HzWPm5xHc2F9LgkXQ+T/21tiGwCdWspzE+i8ZYjjHYPr/8OdSPkE0fRR9l3H9vSGofl26q6RJ+YfuRBkIaeJJeavt/xyqL7gx7f2QSRcQU1Gm8/7DeAzARJP2Eqj9yES39kbaHopaWpqc+yLj/mKwkvYTqOc6z2sb/rwdMayaqoTDU/ZFJFP2xR4eyDYHtJWXcfzRpBrAO1f/91nnH7gX2aSSi4TDU/ZFpeppAGfffO5JeaPuqpuMYVJK2KsOL1wGwfX/TMQ2yYe+PTI1iAmXcf099EUh7+vitK+lqqpoukm4HDrD9s2bDGky2t246hn5KophAGfffU2o6gAF3EvB3ti8DkPTyUjaU02T3y1Tpj0yi6IOM+58QQzGHToPWHkkSALa/I2ntJgMaUFOiPzJ9FH2Qcf8x2Uk6D7gKOKMUvQl4ke3XNRfV8Bi2/sgkiogpSNLTqWplu1D9kPk+8GHbdzUa2BAZpvtS0vQUMcVImgZ8YFgmrJuMhq0/MokiBoqk5wMvK4vft/2TJuMZRLYfk7RL03EMg6nSH5mmpwmScf+rT9KhwDuBc0vR64CTbH+muagGk6TPA5tRTYX/wEi57XNH3SmeYqr0RyZRTJBhaq9sSnmQ/UtsP1CW1wZ+ZHv7ZiMbPKM8NdC23z7hwcSkl6aniZNx/6tPrPwAqMfIdR2v99m+vekgYjAkUUycjPtffacCl5ehnQB7Ayc3GM/AkbQHcArwqKTHgf1s/7DhsGKSS9NTDBRJL6Qa0glVZ/bVTcYzaErz3X62fy5pZ+AE2+3t7LEahrE/MokiBoakfwXOzC/g8WvvK0vfWe8N4zVN01MMkkXAB8sY9fOoksZQP6u4D57R9hyKlZZtf6qBmIbN0PWbpUbRZxn333uSNgTeAMwHtrQ9p+GQBoakY+rW205f2mqStLftbzQdRy8lUfRRxv33h6SdgDcCewHX2+40MVtE9EgSRR9l3H9vSTqBKtn+EjgLOM/23c1GNVgkHWH7BEmf4al3FJNpPaKT9FH0V8b999YvqRJvxv+P3/Xlz/TtRNdSo+ij0kl4AFXHK1Tj/k+z/enmohpskvYE/rQsftf2+U3GEzFimPsjkyj6LOP+e0fSx4GdgC+Xov2BK22/v7moBpOkbYH3ArNpaVmw/YqmYhpkw94fmUTRRxn331ulz2cH24+X5WnA1enzWXWSfgL8O9WQ4yeaR20vaiyoATbs/ZHpo+ivjPvvvQ2AO8v79ZsMZMCtsP35poMYIkPdH5lE0Ue2TwdObxn3f7ykjPsfv48DV0u6jOo/4Z8CRzYb0sA6X9LfUv2AeeIBO7bvHH2XqDHU85Cl6WkCZNx/70jaBNixLF5h+zdNxjOoJN3Yodi2nzXhwQyJYe6PTKLoo4z77z1JmwFbsXIH7Peaiyhi+Psj0/TUXxn330OSjqeqmV0LPF6KDSRRdEnS69uKDNwOLLZ9XwMhDYuh7o9MjaLPMu6/dyQtAba3PTQPrZ9oozzZbkNge+BA25dOcEhDZVjnIUuNoo86jPs/RNJLMu5/3G4A1qSl8zVWje23dSqXtBVwNrDzxEY0dLYBnkvVPHr9GNsOjNQo+ijj/ntL0teB5wOXsPJIncxP1APD+ByFiTLs/ZGpUfRfxv33zoLyih4rbeupqY3fUPdHpkbRR5L2Bz4BrDTu3/ZZjQYWU5ak83nqrLEbApsAb7L9o4mPajgMc39kEkWfZdx/70iaQ3XT3VzgaSPlGfvfPUntz8c2cAfwC9uPNBDSUBj2eciSKPos4/57R9IPgGOAE4E9gLcBa9j+UKOBxZQ37P2R6aPoo4z777mZti+RJNs3A8dKWgQkUcRkMLT9kUkU/bU38JyM+++ZhyWtAfxC0sHAMmCdhmOKgCGfhyxNT30k6VvAvrbvbzqWYSBpR6qx6RsA/wisB5xg+/JGAxtwkl5o+6qm4xh0w9wfmRpFf/0eWCwp4/57Y7btK4H7qfonkLQvkESxer4I5P6J1bcG1XQo04FtJW07LP2RqVH0kaQDOpWX6cdjFXW6ISw3ia0+SVfbfkHTcQyy0fojbe/ZXFS9kxpFHyUh9Iak3YHXAJuVWTpHrAesaCaqofLhpgMYAkPdH5lE0UcZ998ztwELgT2pZukccR9weCMRDRHb32g6hiEw1POQpempjzLuv7ckrUn142ZL20uajidixLDPQ5ZE0UeSFtl+kaRrbP9xa1nTsQ0iSXsAnwRm2N5a0g7AR4alHTgG17D3R6bpqb8y7r+3jqWaJuE7ALYXS9q6yYAGmaTnAy8ri9+3/ZMm4xlkw5IQRrNG0wEMuUOBPwAOAV4EvAl4S6MRDbZHbd/TVpYq8ThIOpRqXqJnlNeXJL2n2agGl6Q5ks6RdJ2kG0ZeTcfVK6lR9FfG/ffWtZL+CphWBgocAgzlM4onwIHAzrYfgCeGd/4I+EyjUQ2uU3myP/LPKf2RjUbUQ0NzIpPUUV2WRXfeA2xH1Vn4VeBe4LBGIxpcAh5rWX6slMX4zLR9CVW/7822jwX+suGYeiY1ij7IuP/+sP174APlFavnVOBySeeV5b2BkxuMZ9ANdX9kRj31Qekk3AH4CCvPbHofcJntuxoJbMBJmge8H5jNytO2D8VUzhNN0guBXcri921f3WQ8g2zY5yFLouijjPvvLUlLgPcB1/DkNAmUKcdjFZSa7pm208fTA5L2tf21scoGVfoo+ms3YDHw3wCSdpCUZz6P33LbC2zfWNqBb06SGLdFwAcl/VLSJ0ttLcZvqPsjU6Poo/JQnVcA3xmZdK315rtYNZJ2pXrEZPvdr+c2FtSAk7Qh8AZgPlXNd07DIQ2Ulv7I/YCzWlatB8y1vVMjgfVYOrP761Hb90grDSZJZh6/twHPpZpTp/WJgUkU47cN1TXdiqqNPVbNlJiHLImivzLuv7d2tP2cpoMYBpJOAF4H/JLql/A/2r672agGT7mb/SeSvsIQ90emj6K/Mu6/t34oaW7TQQyJXwIvsb2b7VOTJFbbUPdHpo8iBoak64FnAzdSJV9RPRwmw2PHQdKeVM92Bviu7fObjGeQDXt/ZJqe+ijj/ntut6YDGBaSPk41weKXS9Ehkl5i+/0NhjXIhro/Momiv75Mh3H/MT4ZCttTfwnsYPtxAEmnA1dT/bCJVTfU/ZHpo+ivjPuPyWyDlvfrNxbFcBjq/sj0UfRRxv3HZCVpf+ATwGVUfT1/Chxp+6zaHWNKSqLoI0lfohqjfi0t4/5tv725qCIqkjYBdiyLV9j+TZPxDLJh749MougjSUsy7j8mK0mbUd1o1/rF9r3mIhpcwz4PWTqz++uHkubavq7pQCJalQcVvZG22i6QRDE+y20PzX0T7VKj6KOM+4/JqvwC3t72w2NuHGMa9v7I1Cj6K+P+Y7K6gWrOrCSK3hjqeciSKPpoWNonYyj9Hlgsqf0X8CHNhTTQhnoesiSKiKlpQXlFbwx1f2T6KCIiVtOw90cmUURMQWWaiY8Dc4GnjZTbflZjQQ0wSVt1Kh+W5uc0PUVMTacCxwAnAn9O1RmbKX3GaVgSwmhSo4iYgiQtsv2i1qmwR8qaji0mn9QoIqamhyWtAfxC0sHAMmCdhmOKSSo1iogpSNKOVM/I3gD4R2A94ATblzcaWExKaZOMmJpm277f9q2232b7DcCWTQcVk1NqFBFTkKSrbL9wrLIISB9FxJQiaXfgNcBmkv61ZdV6wIpmoorJLokiYmq5DVgI7Aksaim/Dzi8kYhi0kvTU8QUJGlNqh+KW9pe0nQ8MbmlMztiatoNWAz8N4CkHSRl7qfoKIkiYmo6FtgJuBvA9mJg6yYDiskriSJianrU9j1tZWmHjo7SmR0xNV0r6a+AaWWCwEOAHzYcU0xSqVFETE3vAbajmhL7q8C9wGGNRhSTVkY9RURErTQ9RUxBkuYB7wdm0/I9MCwP2oneSo0iYgqStAR4H3AN8PhI+bA/VyHGJzWKiKlpue3cNxFdSY0iYgqStCuwP3AJVYc2ALbPbSyomLRSo4iYmt4GPBdYkyebngwkUcRTpEYRMQVJWmL7OU3HEYMh91FETE0/lDS36SBiMKRGETEFSboeeDZwI1UfhQBneGx0kkQRMQVJ2qpTeYbHRidJFBERUSt9FBERUSuJIiIiaiVRRERErSSKiIiolUQRERG1/j8QIY84KfQ1AgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Count</th>\n",
       "      <th>PEP Score - median</th>\n",
       "      <th>PEP Score - std. dev.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Human</th>\n",
       "      <td>2559.0</td>\n",
       "      <td>0.000158</td>\n",
       "      <td>0.011650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Other</th>\n",
       "      <td>27061.0</td>\n",
       "      <td>0.000949</td>\n",
       "      <td>0.017429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Contaminant</th>\n",
       "      <td>254.0</td>\n",
       "      <td>0.001017</td>\n",
       "      <td>0.011370</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Reverse</th>\n",
       "      <td>69.0</td>\n",
       "      <td>0.004650</td>\n",
       "      <td>0.008234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Count  PEP Score - median  \\\n",
       "metanovo - UniProt: Human         2559.0            0.000158   \n",
       "metanovo - UniProt: Other        27061.0            0.000949   \n",
       "metanovo - UniProt: Contaminant    254.0            0.001017   \n",
       "metanovo - UniProt: Reverse         69.0            0.004650   \n",
       "\n",
       "                                 PEP Score - std. dev.  \n",
       "metanovo - UniProt: Human                     0.011650  \n",
       "metanovo - UniProt: Other                     0.017429  \n",
       "metanovo - UniProt: Contaminant               0.011370  \n",
       "metanovo - UniProt: Reverse                   0.008234  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pep_scores = []\n",
    "names = []\n",
    "keys = list(stool_results.keys())\n",
    "for key in keys:\n",
    "    peps = stool_results[key]['TargetPeptides']\n",
    "    cpeps = peps[peps['Potential contaminant']=='+']\n",
    "    peps = peps[peps['Potential contaminant'] != '+']\n",
    "    #peps = peps[peps['PEP'] < 1]\n",
    "    rpeps = stool_results[key]['ReversePeptides']\n",
    "    key_peps = set(peps['Sequence'].tolist()).copy()\n",
    "    non_human= set(key_peps)\n",
    "    human = set(key_peps)\n",
    "    #for qkey in keys:\n",
    "    #|    if not qkey == key:\n",
    "    non_human -= all_human_peps\n",
    "    human -= all_non_human_peps\n",
    "            \n",
    "    print('{} Human: '.format(key), len(human))\n",
    "    print('{} Non-human: '.format(key), len(non_human))\n",
    "        \n",
    "    human_pep = peps[peps['Sequence'].isin(human)]['PEP'].tolist()\n",
    "    pep_scores.append(human_pep)\n",
    "    names.append(key + ': Human')\n",
    "    \n",
    "    non_human_pep = peps[peps['Sequence'].isin(non_human)]['PEP'].tolist()\n",
    "    pep_scores.append(non_human_pep)\n",
    "    names.append(key + ': Other')\n",
    "    \n",
    "    contaminant_pep = cpeps['PEP'].tolist()\n",
    "    pep_scores.append(contaminant_pep)\n",
    "    names.append(key + ': Contaminant' )\n",
    "    \n",
    "    reverse_pep = rpeps['PEP'].tolist()\n",
    "    pep_scores.append(reverse_pep)\n",
    "    names.append(key + ': Reverse')\n",
    "\n",
    "#############\n",
    "## BOXPLOT ##\n",
    "#############\n",
    "\n",
    "colours = ['b','g','r','c','m','y','k']\n",
    "fig = plt.figure()\n",
    "ax = fig.add_subplot(111)\n",
    "# Create the boxplot\n",
    "bp = ax.boxplot(pep_scores, patch_artist=True, showfliers=False)\n",
    "## change outline color, fill color and linewidth of the boxes\n",
    "count = 0\n",
    "col_ind=0\n",
    "for box in bp['boxes']:\n",
    "    count += 1\n",
    "    # change outline color\n",
    "    box.set( color='#7570b3', linewidth=1)\n",
    "    # change fill color\n",
    "    box.set( facecolor = colours[col_ind] )\n",
    "    ## change color and linewidth of the whiskers\n",
    "    if count % 4 == 0:\n",
    "        col_ind +=1\n",
    "for whisker in bp['whiskers']:\n",
    "    whisker.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the caps\n",
    "for cap in bp['caps']:\n",
    "    cap.set(color='#7570b3', linewidth=1)\n",
    "## change color and linewidth of the medians\n",
    "for median in bp['medians']:\n",
    "    median.set(color='#b2df8a', linewidth=2)\n",
    "    #median.set(linewidth=2)\n",
    "## change the style of fliers and their fill\n",
    "for flier in bp['fliers']:\n",
    "    flier.set(marker='.', color='#e7298a', alpha=0.5)\n",
    "## Custom x-axis labels\n",
    "ax.set_xticklabels(names, rotation=90)\n",
    "#ax.set_yticklabels('Posterior Error Probability (PEP)') \n",
    "ax.set_title('Peptide PEP Score distributions')\n",
    "## Remove top axes and right axes ticks\n",
    "ax.get_xaxis().tick_bottom()\n",
    "ax.get_yaxis().tick_left()\n",
    "fig.savefig('analysis/psm_pep_scores_stool.png', bbox_inches='tight')\n",
    "plt.show()\n",
    "fig.clf()\n",
    "\n",
    "pep_df = pd.DataFrame()\n",
    "for val in range(len(names)):\n",
    "    pep_df.loc[names[val], 'Count'] = len(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - median'] = np.median(pep_scores[val])\n",
    "    pep_df.loc[names[val], 'PEP Score - std. dev.'] = np.std(pep_scores[val])\n",
    "\n",
    "pep_df.to_csv('analysis/stool_pep_scores.csv')\n",
    "pep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KruskalResult(statistic=261.7432734591897, pvalue=1.8866788880406473e-56)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>metanovo - UniProt: Human</th>\n",
       "      <th>metanovo - UniProt: Other</th>\n",
       "      <th>metanovo - UniProt: Contaminant</th>\n",
       "      <th>metanovo - UniProt: Reverse</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Human</th>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>7.074290e-51</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>3.375956e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Other</th>\n",
       "      <td>7.074290e-51</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "      <td>0.994544</td>\n",
       "      <td>6.978041e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Contaminant</th>\n",
       "      <td>2.475835e-06</td>\n",
       "      <td>9.945444e-01</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>1.437990e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>metanovo - UniProt: Reverse</th>\n",
       "      <td>3.375956e-15</td>\n",
       "      <td>6.978041e-08</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>-1.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 metanovo - UniProt: Human  \\\n",
       "metanovo - UniProt: Human                    -1.000000e+00   \n",
       "metanovo - UniProt: Other                     7.074290e-51   \n",
       "metanovo - UniProt: Contaminant               2.475835e-06   \n",
       "metanovo - UniProt: Reverse                   3.375956e-15   \n",
       "\n",
       "                                 metanovo - UniProt: Other  \\\n",
       "metanovo - UniProt: Human                     7.074290e-51   \n",
       "metanovo - UniProt: Other                    -1.000000e+00   \n",
       "metanovo - UniProt: Contaminant               9.945444e-01   \n",
       "metanovo - UniProt: Reverse                   6.978041e-08   \n",
       "\n",
       "                                 metanovo - UniProt: Contaminant  \\\n",
       "metanovo - UniProt: Human                               0.000002   \n",
       "metanovo - UniProt: Other                               0.994544   \n",
       "metanovo - UniProt: Contaminant                        -1.000000   \n",
       "metanovo - UniProt: Reverse                             0.000001   \n",
       "\n",
       "                                 metanovo - UniProt: Reverse  \n",
       "metanovo - UniProt: Human                       3.375956e-15  \n",
       "metanovo - UniProt: Other                       6.978041e-08  \n",
       "metanovo - UniProt: Contaminant                 1.437990e-06  \n",
       "metanovo - UniProt: Reverse                    -1.000000e+00  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "post_hoc = list_kw_dunn(names, pep_scores, \"PEP\", \"Workflow\", 'analysis')\n",
    "post_hoc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
